core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -0,0 +1,919 @@
+        if (validateOffsetMetadataLength(offsetAndMetadata.metadata))
+          (topicAndPartition, responseCode)
+        else
+          (topicAndPartition, ErrorMapping.OffsetMetadataTooLargeCode)
+      }
+
+      // finally trigger the callback logic passed from the API layer
+      responseCallback(commitStatus)
+    }
+
+    // call replica manager to append the offset messages
+    replicaManager.appendMessages(
+      config.offsetCommitTimeoutMs.toLong,
+      config.offsetCommitRequiredAcks,
+      true, // allow appending to internal offset topic
+      offsetsAndMetadataMessageSet,
+      putCacheCallback)
+  }
+
+  /**
+   * The most important guarantee that this API provides is that it should never return a stale offset. i.e., it either
+   * returns the current offset or it begins to sync the cache from the log (and returns an error code).
+   */
+  def getOffsets(group: String, topicPartitions: Seq[TopicAndPartition]): Map[TopicAndPartition, OffsetMetadataAndError] = {
+    trace("Getting offsets %s for group %s.".format(topicPartitions, group))
+
+    val offsetsPartition = partitionFor(group)
+
+    /**
+     * we need to put the leader-is-local check inside the offsetLock to protects against fetching from an empty/cleared
+     * offset cache (i.e., cleared due to a leader->follower transition right after the check and clear the cache).
+     */
+    inReadLock(offsetRemoveLock) {
+      if (partitionLeaderIsLocal(offsetsPartition)) {
+        if (topicPartitions.isEmpty) {
+          // Return offsets for all partitions owned by this consumer group. (this only applies to consumers that commit offsets to Kafka.)
+          offsetsCache.filter(_._1.group == group).map { case(groupTopicPartition, offsetAndMetadata) =>
+            (groupTopicPartition.topicPartition, OffsetMetadataAndError(offsetAndMetadata.offset, offsetAndMetadata.metadata, ErrorMapping.NoError))
+          }.toMap
+        } else {
+          topicPartitions.map { topicAndPartition =>
+            val groupTopicPartition = GroupTopicPartition(group, topicAndPartition)
+            (groupTopicPartition.topicPartition, getOffset(groupTopicPartition))
+          }.toMap
+        }
+      } else {
+        debug("Could not fetch offsets for group %s (not offset coordinator).".format(group))
+        topicPartitions.map { topicAndPartition =>
+          val groupTopicPartition = GroupTopicPartition(group, topicAndPartition)
+          (groupTopicPartition.topicPartition, OffsetMetadataAndError.NotCoordinatorForGroup)
+        }.toMap
+      }
+    }
+  }
+
+  /**
+   * Asynchronously read the partition from the offsets topic and populate the cache
+   */
+  def loadGroupsForPartition(offsetsPartition: Int) {
+
+    val topicPartition = TopicAndPartition(GroupCoordinator.GroupMetadataTopicName, offsetsPartition)
+
+    loadingPartitions synchronized {
+      ownedPartitions.add(offsetsPartition)
```

Comment:
I think this does not make a major difference here: with this approach when a partition is loading, coordinator responses will get GROUP_LOADING error code causing the client to retry; with the suggested approach the coordinator responses will get NOT_COORDINATOR_FOR_GROUP error code causing the client to re-discover and retry. I personally prefer the first approach as it does not introduce necessary re-discovering.